src/CorePolicy.cc

```diff
@@ -0,0 +1,201 @@
+  */
+double *utilizationThresholds;
+
+/**
+  * The difference in load, expressed as a fraction of a core, between a
+  * ramp-down threshold and the corresponding ramp-up threshold.
+  */
+double idleCoreFractionHysteresis = 0.2;
+
+/**
+  * The percentage of slots which are occupied that will prevent reducing the
+  * number of cores.
+  */
+double SLOT_OCCUPANCY_THRESHOLD = 0.5;
+
+/**
+  * The period in ns over which we measure before deciding to reduce the number
+  * of cores we use.
+  */
+const uint64_t MEASUREMENT_PERIOD = 50 * 1000 * 1000;
+
+void coreLoadEstimator();
+
+/**
+  * Bootstrap the core load estimator thread.  This must be done separately
+  * from the CorePolicy constructor because the constructor must run before
+  * Arachne adds any cores, but the core load estimator must run after
+  * Arachne has cores.
+  */
+void CorePolicy::bootstrapLoadEstimator(bool disableLoadEstimation) {
+    if (!disableLoadEstimation) {
+        utilizationThresholds = new double[Arachne::maxNumCores];
+        Arachne::createThread(baseClass, coreLoadEstimator);
+    }
+}
+
+/**
+  * Choose a core to deschedule.  Return its coreId.
+  */
+int CorePolicy::chooseRemovableCore() {
+    threadCoreMapEntry* entry = threadCoreMap[baseClass];
+    uint8_t minLoaded =
+        Arachne::occupiedAndCount[entry->map[0]]->load().numOccupied;
+    int minCoreId = entry->map[0];
+    for (uint32_t i = 1; i < threadCoreMap[baseClass]->numFilled; i++) {
+        uint32_t coreId = entry->map[i];
+        if (Arachne::occupiedAndCount[coreId]->load().numOccupied < minLoaded) {
+            minLoaded = Arachne::occupiedAndCount[coreId]->load().numOccupied;
+            minCoreId = coreId;
+        }
+    }
+    return minCoreId;
+}
+
+/**
+  * Update threadCoreMap when a new core is added.  Takes in the coreId
+  * of the new core.
+**/
+void CorePolicy::addCore(int coreId) {
+  threadCoreMapEntry* entry = threadCoreMap[baseClass];
+  entry->map[entry->numFilled] = coreId;
+  entry->numFilled++;
+}
+
+/**
+  * Update threadCoreMap when a core is removed.  Takes in the coreId
+  * of the doomed core.
+**/
+void CorePolicy::removeCore(int coreId) {
+  threadCoreMapEntry* entry = threadCoreMap[baseClass];
+  entry->numFilled--;
+  int saveCoreId = entry->map[entry->numFilled];
+  for (uint32_t i = 0; i < entry->numFilled; i++) {
+    if (entry->map[i] == coreId) {
+      entry->map[i] = saveCoreId;
+      return;
+    }
+  }
+}
+
+threadCoreMapEntry* CorePolicy::getThreadCoreMapEntry(threadClass_t threadClass) {
+  return threadCoreMap[threadClass];
+}
+
+/**
+  * Periodically wake up and observe the current load in Arachne to determine
+  * whether it is necessary to increase or reduce the number of cores used by
+  * Arachne.
+  */
+void coreLoadEstimator() {
+    Arachne::PerfStats previousStats;
+    Arachne::PerfStats::collectStats(&previousStats);
+
+    for (Arachne::PerfStats currentStats; ; previousStats = currentStats) {

```

Comment:
What does one iteration of this loop correspond to? That would make for a good comment just above the loop.